integration/framework/service.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/expfmt"
+)
+
+var (
+	dockerPortPattern = regexp.MustCompile(`^.*:(\d+)$`)
+)
+
+type Service struct {
+	name         string
+	image        string
+	networkName  string
+	networkPorts []int
+	env          map[string]string
+	command      *Command
+	readiness    *ReadinessProbe
+
+	// Maps container ports to dynamically binded local ports.
+	networkPortsContainerToLocal map[int]int
+
+	// Generic retry backoff.
+	retryBackoff *util.Backoff
+}
+
+func NewService(
+	name string,
+	image string,
+	networkName string,
+	networkPorts []int,
+	env map[string]string,
+	command *Command,
+	readiness *ReadinessProbe,
+) *Service {
+	return &Service{
+		name:                         name,
+		image:                        image,
+		networkName:                  networkName,
+		networkPorts:                 networkPorts,
+		env:                          env,
+		command:                      command,
+		networkPortsContainerToLocal: map[int]int{},
+		readiness:                    readiness,
+		retryBackoff: util.NewBackoff(context.Background(), util.BackoffConfig{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 200 * time.Millisecond,
+			MaxRetries: 100, // Sometimes the CI is slow
+		}),
+	}
+}
+
+func (s *Service) Start() error {
+	cmd := exec.Command("docker", s.buildDockerRunArgs()...)
+	cmd.Stdout = &LinePrefixWriter{prefix: s.name + ": ", wrapped: os.Stdout}
+	cmd.Stderr = &LinePrefixWriter{prefix: s.name + ": ", wrapped: os.Stderr}
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// Wait until the container has been started
+	if err := s.WaitStarted(); err != nil {
+		return err
+	}
+
+	// Get the dynamic local ports mapped to the container
+	for _, containerPort := range s.networkPorts {
+		out, err := RunCommandAndGetOutput("docker", "port", s.name, strconv.Itoa(containerPort))
+		if err != nil {
+			return errors.Wrapf(err, "unable to get mapping for port %d", containerPort)
+		}
+
+		stdout := strings.TrimSpace(string(out))
+		matches := dockerPortPattern.FindStringSubmatch(stdout)
+		if len(matches) != 2 {
+			return fmt.Errorf("unable to get mapping for port %d (output: %s)", containerPort, stdout)
+		}
+
+		localPort, err := strconv.Atoi(matches[1])
+		if err != nil {
+			return errors.Wrapf(err, "unable to get mapping for port %d", containerPort)
+		}
+
+		s.networkPortsContainerToLocal[containerPort] = localPort
+	}
+
+	return nil
+}
+
+func (s *Service) Stop() error {
+	fmt.Println("Stopping", s.name)
+
+	if out, err := RunCommandAndGetOutput("docker", "stop", "--time=30", s.name); err != nil {
+		fmt.Println(string(out))
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) Kill() error {
+	fmt.Println("Killing", s.name)
+
+	if out, err := RunCommandAndGetOutput("docker", "stop", "--time=0", s.name); err != nil {
+		fmt.Println(string(out))
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) Endpoint(port int) string {
```

Comment:
This `port` parameter is also unused.

Reply:
Damn. Thanks! You spotted another issue. It should have been `s.networkPortsContainerToLocal[port]` (this is the result of a bunch of refactories I did while building it).